### PR TITLE
refactor: error reporting to drive category from stats

### PR DESCRIPTION
**Pull request by koladilip**
# Description

Propagate statTags from transformer response to reporting.
Address sonar issues of too many function params and reduced code complexity

## Linear Ticket

https://linear.app/rudderstack/issue/INT-2973/[chore]-propagate-error-category-from-stattags-of-transformer
Resolves INT-2973

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: enterprise/reporting/error_extractor.go [hunk 1]
```diff
@@ -109,48 +109,77 @@ func (ext *ExtractorHandle) getSimpleMessage(jsonStr string) string {
...
+}
+
+func (ext *ExtractorHandle) handleKey(key string, value interface{}) *string {
+	valueStr, isString := value.(string)
+	switch key {
+	case "reason", "Error", responseKey, errorKey:
 		if !isString {
-			ext.log.Debugf("Type-assertion failed for %v with value %v: not a string", key, erRes)
+			ext.log.Debugf("Expected string value for key: %v, but got: %T", key, value)
+			return nil
```
Comment (achettyiitr) [inline enterprise/reporting/error_extractor.go:125]:
```suggestion
			ext.log.Debugn("Handling key", logger.NewStringField("key", key), logger.NewField("value", value))
			return nil
```

File: enterprise/reporting/error_extractor.go [hunk 2]
```diff
@@ -109,48 +109,77 @@ func (ext *ExtractorHandle) getSimpleMessage(jsonStr string) string {
...
+	if len(valueStr) == 0 {
+		return nil
+	}
+	return &valueStr
+}
+
+func (ext *ExtractorHandle) handleWarehouseError(value interface{}, key string) *string {
+	valAsMap, isMap := value.(map[string]interface{})
+	if !isMap {
+		ext.log.Debugf("Failed type assertion to map[string]interface{} for warehouse error key: %s", key)
```
Comment (achettyiitr) [inline enterprise/reporting/error_extractor.go:178]:
```suggestion
		ext.log.Debugn("Failed type assertion to map[string]interface{} for warehouse error key", logger.NewStringField("key", key), logger.NewField("value", value))
```

File: enterprise/reporting/error_extractor.go [hunk 3]
```diff
@@ -109,48 +109,77 @@ func (ext *ExtractorHandle) getSimpleMessage(jsonStr string) string {
...
+			return *result
+		}
+	}
+	return ""
+}
+
+func (ext *ExtractorHandle) handleKey(key string, value interface{}) *string {
+	valueStr, isString := value.(string)
+	switch key {
+	case "reason", "Error", responseKey, errorKey:
```
Comment (achettyiitr) [inline enterprise/reporting/error_extractor.go:122]: We can move `value.(string)` within the case block. 
```suggestion
func (ext *ExtractorHandle) handleKey(key string, value interface{}) *string {
	switch key {
	case "reason", "Error", responseKey, errorKey:
		valueStr, isString := value.(string)
```

File: enterprise/reporting/error_extractor.go [hunk 4]
```diff
@@ -109,48 +109,77 @@ func (ext *ExtractorHandle) getSimpleMessage(jsonStr string) string {
...
-		erResStr, isString := erRes.(string)
+		if result := ext.handleKey(key, erRes); result != nil {
+			return *result
+		}
+	}
+	return ""
+}
+
+func (ext *ExtractorHandle) handleKey(key string, value interface{}) *string {
+	valueStr, isString := value.(string)
```
Comment (achettyiitr) [inline enterprise/reporting/error_extractor.go:120]: GO's ok idiom. https://www.devzery.com/post/mastering-ok-in-golang-your-ultimate-guide
```suggestion
	valueStr, ok := value.(string)
```

**Review by achettyiitr — APPROVED** [review]

File: enterprise/reporting/error_extractor.go [hunk 5]
```diff
@@ -109,48 +109,77 @@ func (ext *ExtractorHandle) getSimpleMessage(jsonStr string) string {
 	}
 
 	for key, erRes := range jsonMap {
-		erResStr, isString := erRes.(string)
+		if result := ext.handleKey(key, erRes); result != nil {
```
Comment (achettyiitr) [inline enterprise/reporting/error_extractor.go:112]: Didn't get the motivation behind differentiating empty string vs no string. Was it a bug earlier?

Comment (koladilip) [reply enterprise/reporting/error_extractor.go]: addressed

File: enterprise/reporting/error_extractor.go (hunk 4 above)
Comment (koladilip) [reply enterprise/reporting/error_extractor.go]: addressed

**Review by sanpj2292 — APPROVED** [review]

**Pull request merged**